README.md
# HistogramZoo

1. [Description](#description)
2. [Installation](#installation)
3. [Quick start](#quick-start)
4. [Resources](#resources)
5. [Getting help](#getting-help)
6. [Citation information](#citation-information)
7. [License](#license)

## Description
HistogramZoo is a generalized framework for histogram segmentation and statistical characterization written in R. Histograms are represented as S3 Histogram or GenomicHistogram objects. For the characterization of genomic data on a linear coordinate system, such as the human reference genome, HistogramZoo provides several convenient data loaders to compute coverage of BED files and load bigWig files, and support both genomic and transcriptomic coordinate systems.

The characteristic function provided by HistogramZoo is `segment_and_fit` which is a wrapper for a sequence of operations which aim to segment complicated histograms and identify regions of enriched density and fit known statistical distributions to these regions. Using the Fine-to-Coarse segmentation algorithm described in [Delon *et al.,* 2006](https://link.springer.com/chapter/10.1007/11492542_30), segmentation occurs iteratively on a set of local optima. The user is provided with the option of filtering low entropy regions, fitting truncated distributions, fine-tuning the endpoints of uniformly-distributed regions and using several goodness-of-fit metrics to improve reported results.

Finally, HistogramZoo provides several convenient functions to summarize fits and visualize data.

![overview plot](man/figures/Overview.png)

## Installation

Using devtools in R:
```R
library(devtools);
install_github('https://github.com/uclahs-cds/public-R-HistogramZoo'); 
```

From source:
```shell script
git clone https://github.com/uclahs-cds/public-R-HistogramZoo.git
R CMD INSTALL public-R-HistogramZoo
```

From CRAN (coming soon!)

## Quick start

A basic example of applying HistogramZoo to a histogram of Gaussian data is provided below:

```R

library(HistogramZoo);

set.seed(271828);

# Generating Data
gaussian_data <- rnorm(10000, mean = 50, sd = 15);
histogram_data <- observations_to_histogram(gaussian_data, histogram_bin_width=5);

# Segmentation and fitting distributions
results <- segment_and_fit(histogram_data, eps = 1);

# Summarizing results
results_table <- summarize_results(results);

# Plotting
create_coverageplot(results);

```

**Results table**

| region_id | segment_id | start | end | interval_count | interval_sizes | interval_starts | histogram_start | histogram_end | value        | metric    | dist | dist_param1 | dist_param2 | dist_param1_name | dist_param2_name |
|-----------|------------|-------|-----|----------------|----------------|-----------------|-----------------|---------------|--------------|-----------|------|-------------|-------------|------------------|------------------|
| -21-102   |          1 |    19 |  79 |              1 | 61,            | 1,              |               9 |            20 | 0.9554924687 | consensus | norm | 33.03241096 | 14.88105209 | mean             | sd               |

**Coverage plot**
![coverage plot](man/figures/hz_output.png)

## Resources

- [Function reference](https://uclahs-cds.github.io/public-R-HistogramZoo/reference/index.html)
- [Vignette](https://uclahs-cds.github.io/public-R-HistogramZoo/articles/HistogramZoo.html)

## Getting help

Looking for guidance or support with HistogramZoo? Look no further.

* Check out our [Discussions](https://github.com/uclahs-cds/public-R-HistogramZoo/discussions) page!
* Submit bugs :bug:, suggest new features :cherry_blossom: or see current work :mechanical_arm: at our [Issues](https://github.com/uclahs-cds/public-R-HistogramZoo/issues) page.

## Citation information

You have stumbled upon an unpublished software :shushing_face: :shushing_face: :shushing_face:. We are currently preparing the manuscript for HistogramZoo. Please befriend us to learn more or check back later for updated citation information.

## License

Authors: Helen Zhu, Stefan Eng, Paul C. Boutros (PBoutros@mednet.ucla.edu)

HistogramZoo is licensed under the GNU General Public License version 3.0. See the file LICENSE.md for the terms of the GNU GPL license.

HistogramZoo is a generalized framework for histogram segmentation and statistical characterization written in R.

Copyright (C) University of California Los Angeles ("Boutros Lab") All rights reserved.

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.


NEWS.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

# HistogramZoo v1.5.2

## [1.5.2] - 2023-08-24

### Added
- more inherited parameters for plotting code
- number of local optima to returned results
- documentation for simulation function

### Fixed
- bugs in simulation code - seed placement for reproducibility
- Closes #149 - fixed bug with coverageplot legend colors

## [1.5.1] - 2023-08-02

### Added
- Analysis for multi-peak simulations

## [1.5.0] - 2023-07-26

### Added
- Uni- and Multi-peak simulations
- Plots for unimodal simulations

## [1.4.2] - 2023-07-03

### Added
- Add `+.Histogram` support to add bin width 1 histograms together.

## [1.4.1] - 2023-05-02

### Added
- ModelFit class to allow pretty-printing of `fit_distributions` results

### Fixed
- Bugs associated with fitting truncated distributions

## [1.4.0] - 2023-04-19

### Added
- S3 methods to allow numeric/Histogram/GenomicHistogram input in key segment_and_fit functions [Pull #121](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/121)
  - consecutive_start and consecutive_end for GenomicHistogram to keep track of disjoint intervals
  - shifted gamma function to allow for non-zero starts

## Changed
- Added summary statistics estimation for GenomicHistogram objects [Pull #124](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/124)
   - Added `bin_width` and `midpoint` estimation for Histogram and GenomicHistogram objects
- Enforcing continuity of bins in Histogram and GenomicHistogram objects when subsetting [Pull #125](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/125)

## Fixed
- Fix vignette/README/image file paths for pkgdown [Pull #115](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/115)

# HistogramZoo v1.3.0

## Added
- Added summary statistics (mean, sd, var, skew) estimation from histogram objects [Pull #113](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/113)

## Changed
- Enforcing bin width and addition of intron-traversing bins in GenomicHistogram [Pull #108](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/108)

# HistogramZoo v1.2.0

## Added
- Added a classic histogram plotting function [Pull #111](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/111)

# HistogramZoo v1.1.0

## Added
- Maxiumum likelihood estimation for distribution parameters [Pull #97](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/97)
- A vignette! [Pull #81](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/81)
- Added `gamma_flip` as an available distribution [Pull #62](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/62)

## Changed
- New rules for `Histogram` and `GenomicHistogram` [Pull #76](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/76)
  - `Histogram` uses a strictly continuous Base 0 system while `GenomicHistogram` uses a potentially disjoint Base 1 system
  - Both are restricted to having a fixed `bin_width` in (n-1) bins
- Add pkgdown build on new release [Pull #56](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/56)
- Updated `find_consensus_model` to include user-specified weights and metric prioritization & addition of new RRA method [Pull #53](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/53)

## Fixed
- Bug in `segment_and_fit` - finer segmentation is required to initialize `meaningful_gaps_local` [Pull #73](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/73)
- Refactored `create_trackplot` to use `BoutrosLab.plotting.general::create.scatterplot` as a base function thus increasing efficiency [Pull #71](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/71)
- Refactored `find_uniform_segment` to reduce code redundancy [Pull #59](https://github.com/uclahs-cds/public-R-HistogramZoo/pull/59)

# HistogramZoo v1.0.0

## Added
- First release!


R/bpg.plot.R

bpg.plot = function(
  output.tag,
  output.dir,
  distr.plotting.data,
  geneinfo,
  bin.counts,
  seg.gr,
  p,
  fitted.seg.gr,
  results,
  histogram.metric = c("jaccard", "intersection", "ks", "mse", "chisq")
) {
  histogram.metric = match.arg(histogram.metric)

  metric.label = switch(histogram.metric,
                        "jaccard" = "Jaccard Index",
                        "intersection" = "Histogram Intersection",
                        "ks" = "KS Statistic",
                        "mse" = "Mean Squared Error",
                        "chisq" = "Chi-square statistic"
                        )

  # Plotting data
  n.distributions = length(distr.plotting.data)
  for(i in 1:n.distributions){distr.plotting.data[[i]]$'i' <- i}
  distplot.data = do.call(rbind, distr.plotting.data)

  # Bin counts
  bin.counts = bin.counts[,c("start", "Coverage")]
  bin.counts$dist = "coverage"
  bin.counts$i = n.distributions + 1
  colnames(bin.counts) = c('x', 'dens', 'dist', 'i')

  # Points
  points = bin.counts[bin.counts$x %in% p,]

  # Plotting Data
  lineplot.data = rbind(distplot.data, bin.counts)

  # Residuals
  residual.data = merge(distplot.data[,c("x", "dens")], bin.counts[,c("x", "dens")], by = "x", all = T)
  colnames(residual.data) = c("x", "fitted", "real")
  residual.data$resid = residual.data$real - residual.data$fitted

  # Colours
  col.reference = structure(c("black", "darkorange", "chartreuse4", "darkorchid4"),
                            names = c("coverage", "norm", "gamma", "unif"))
  col.numeric = structure(1:3, names = c("norm", "gamma", "unif"))
  lwd.reference = structure(c(1, 2.5, 2.5, 2.5), names = c("coverage", "norm", "gamma", "unif"))
  col.used = unique(lineplot.data[,c("dist", "i")])
  col.used = col.used[order(col.used$i),]
  col.vec = col.reference[col.used$dist]
  lwd.vec = lwd.reference[col.used$dist]

  # Plotting
  sc = BoutrosLab.plotting.general::create.scatterplot(
    dens ~ x,
    lineplot.data,
    # Groups
    groups = lineplot.data$i,
    col = col.vec,
    # Axes
    xlimits = c(0, geneinfo$exome_length),
    xaxis.cex = 0,
    xlab.cex = 0,
    ylab.cex = 1,
    xaxis.tck = 0,
    yaxis.cex = 1,
    ylab.label = "Coverage (at BP resolution)",
    main.cex = 0,
    # Lines & PCH
    type = c('a'),
    lwd = lwd.vec,
    # Adding extra points
    add.points = T,
    points.x = points$x,
    points.y = points$dens,
    points.pch = 19,
    points.col = 'red'
  )

  resid.chgpts = which(abs(diff(sign(residual.data$resid))) == 2)
  sc.res =  BoutrosLab.plotting.general::create.scatterplot(
    resid ~ x,
    residual.data,
    # Groups
    # groups = residual.data$something,
    col = "black",
    # Axes
    xlimits = c(0, geneinfo$exome_length),
    xaxis.cex = 0,
    xlab.cex = 0,
    ylab.cex = 1,
    xaxis.tck = 0,
    yaxis.cex = 1,
    ylab.label = "Residuals",
    main.cex = 0,
    # Lines & PCH
    type = c('p'),
    cex = 0.01,
    # Adding lines at changepoints
    abline.v = resid.chgpts,
    abline.h = 0,
    abline.lty = "dotted",
    abline.col = "lightgrey",
    abline.lwd = 0.01
  )

  # Generating Gene Segments
  genegr = GenomicRanges::GRanges(seqnames = geneinfo$chr, IRanges::IRanges(start = 1, end = geneinfo$exome_length), strand = geneinfo$strand)
  anno = geneinfo$anno
  anno$start = geneinfo$DNA2RNA[anno$start - geneinfo$left+1]
  anno$stop = geneinfo$DNA2RNA[anno$stop - geneinfo$left+1]
  anno.gr = GenomicRanges::makeGRangesFromDataFrame(anno, keep.extra.columns = T)
  anno.gr = S4Vectors::split(anno.gr, anno.gr$transcript)
  anno.coverage = lapply(anno.gr, GenomicRanges::coverage)
  bins = unlist(GenomicRanges::tile(genegr, width = 1))
  anno.counts = lapply(anno.coverage, function(i) GenomicRanges::binnedAverage(bins, i, "Coverage"))
  transcript.coverage = do.call(cbind, lapply(anno.counts, function(x) x$Coverage))
  transcript.coverage = as.data.frame(transcript.coverage, stringsAsFactors = F)

  # Adding p-values & segments
  # Change seg.gr to fitted.seg.gr for just the fitted segments
  ovl = GenomicRanges::findOverlaps(bins, seg.gr)
  bins$metric = NA
  bins$metric[S4Vectors::queryHits(ovl)] = results$metric[S4Vectors::subjectHits(ovl)]

  # Adding segment fits
  ovl = GenomicRanges::findOverlaps(bins, fitted.seg.gr)
  bins$fitted = 0
  bins$fitted[S4Vectors::queryHits(ovl)] = col.numeric[results$dist[S4Vectors::subjectHits(ovl)]]

  # Plotting Heatmap
  heatmap.data = data.frame(bins, stringsAsFactors = F)
  heatmap.data = heatmap.data[,c("start", "fitted", "metric")]
  heatmap.data = heatmap.data[order(heatmap.data$start),]

  hm.fu = BoutrosLab.plotting.general::create.heatmap(
    heatmap.data[,c("fitted", "fitted")],
    clustering.method = 'none',
    # Plotting Characteristics
    axes.lwd = 1,
    yaxis.tck = 0,
    # Discrete Colours
    at = seq(-0.5, length(col.reference), 1),
    total.colours = length(col.reference) + 1,
    colour.scheme = c('white', col.reference[2:length(col.reference)]),
    # Adding lines for segments
    force.grid.col = TRUE,
    grid.col = TRUE,
    col.lines = c(GenomicRanges::start(fitted.seg.gr), GenomicRanges::end(fitted.seg.gr)),
    # Colourkey
    # colourkey.labels.at = seq(0, length(seg.gr)+1, 1),
    # colourkey.labels = seq(0, length(seg.gr)+1, 1),
    print.colour.key = F
  )

  hm.gof.colour.scheme = c('dodgerblue4', 'firebrick')
  if(histogram.metric %in% c("jaccard", "intersection")) {
    max.at = 1
    min.at = 0
    metric.at =  seq(0.75, 1, 0.05) # fix this when scaled
  } else {
    min.at = 0
    max.at = max(heatmap.data$metric, na.rm = T)
    hm.gof.colour.scheme = rev(hm.gof.colour.scheme)
  }
  metric.at = seq(min.at, max.at, length.out = 21)

  hm.gof = BoutrosLab.plotting.general::create.heatmap(
    heatmap.data[,c("metric", "metric")],
    clustering.method = 'none',
    # Plotting Characteristics
    axes.lwd = 0,
    yaxis.tck = 0,
    # Colours
    colour.scheme = hm.gof.colour.scheme,
    at = metric.at,
    fill.colour = "white",
    # Adding lines for segments
    force.grid.col = TRUE,
    grid.col = TRUE,
    col.lines = c(GenomicRanges::start(seg.gr), GenomicRanges::end(seg.gr)),
    col.lwd = 1,
    # Colourkey
    print.colour.key = F
  )

  x.digits = floor(log10(geneinfo$exome_length))-1
  x.at = seq(0, round(geneinfo$exome_length, digits = -x.digits), length.out = 5)
  hm.coverage = BoutrosLab.plotting.general::create.heatmap(
    transcript.coverage,
    clustering.method = 'none',
    # Axes labels
    xaxis.lab = x.at,
    xat = x.at,
    xaxis.rot = 0,
    xaxis.cex = 1,
    xaxis.tck = 1,
    yaxis.tck = 0,
    xlab.cex = 1,
    xlab.label = "Transcript Coordinate",
    # Discrete Colours
    at = c(-0.5, 0.5, 1.5),
    total.colours = 3,
    colour.scheme = c('white', 'pink'),
    # Border
    axes.lwd = 1,
    # Adding lines for segments
    force.grid.col = TRUE,
    grid.col = TRUE,
    col.lines = c(GenomicRanges::start(seg.gr), GenomicRanges::end(seg.gr)),
    col.lwd = 1,
    force.grid.row = TRUE,
    grid.row = TRUE,
    row.lines = 1:ncol(transcript.coverage)+0.5,
    row.lwd = 1,
    # Colourkey
    print.colour.key = F
  )

  # Legend
  covariate.legend <- list(
    legend = list(
      colours = 'red',
      labels = c("FTC Endpoints"),
      title = expression(bold(underline('Points'))),
      lwd = 0.5
    ),
    legend = list(
      colours = 'black',
      labels = 'Peak Coverage',
      title = expression(bold(underline('Lines'))),
      lwd = 0.5
    ),
    legend = list(
      colours = c('black', 'lightgrey'),
      labels = c('Obs - Exp', 'Intersect points'),
      title = expression(bold(underline('Residuals'))),
      lwd = 0.5
    ),
    legend = list(
      colours = col.reference[2:length(col.reference)],
      labels = names(col.reference)[2:length(col.reference)],
      title = expression(bold(underline('Distributions'))),
      lwd = 0.5
    ),
    legend = list(
      colours = hm.gof.colour.scheme,
      labels = c(min.at, formatC(max.at, format = "e", digits = 2)),
      title = bquote(bold(underline(.(metric.label)))),
      continuous = TRUE,
      height = 2,
      angle = -90,
      tck = 1,
      tck.number = 3,
      at = c(0,100),
      labels.rot = 0,
      lwd = 0.5
    ),
    legend = list(
      colours = c("pink"),
      labels = c("Exon"),
      title = expression(bold(underline('Transcript Coverage'))),
      lwd = 0.5
    )
  )

  side.legend <- BoutrosLab.plotting.general::legend.grob(
    legends = covariate.legend,
    label.cex = 0.7,
    title.cex = 0.7,
    title.just = 'left',
    title.fontface = 'bold',
    size = 2
  )

  # Plotting
  filename = file.path(output.dir, paste0(geneinfo$gene, ".", output.tag, ".SegmentAndFit.pdf"))
  pdf(filename, width = 10, height = 10)

  transcript.height = min(3, ncol(transcript.coverage)*0.5) + 0.55

  mpp = BoutrosLab.plotting.general::create.multipanelplot(
    plot.objects = list(sc, sc.res, hm.fu, hm.gof, hm.coverage),
    plot.objects.heights = c(10, 3, 1, 1, transcript.height),
    y.spacing = -4,
    # Labels
    main = geneinfo$gene,
    main.cex = 2,
    # Legend
    legend = list(
      right = list(
        x = 0.8,
        y = 1,
        fun = side.legend
      )
    )
  )

  print(mpp)
  dev.off()

}
